Transform plugin not called when running with multiple jobs

# Summary
When adding a plugin which registers a transform it registers and calls the transform as expected, but when enabling multiple jobs, the transform doesn't get called resulting in a different result than expected.

The plugin seems to be loaded fine, but the transform is not triggered.

The expected behavior is to have the linting result be the same regardless of the number of jobs.

# Pylint version
```console
> python -m pylint --version
pylint 2.8.3
astroid 2.5.6
Python 3.8.1 (tags/v3.8.1:1b293b6, Dec 18 2019, 23:11:46) [MSC v.1916 64 bit (AMD64)]
```

# MWE
With the content of `mwe.py` being:
```python
# PLUGIN
import astroid


def register(_):
    # Needed for registering the plugin.
    pass


def has_extra_arg_decorator(node):
    return node.decorators and any(d.name == 'class_wrapper' for d in node.decorators.nodes)


def add_extra_arg(node):
    print('transform triggered!')
    init_function = node.getattr('__init__')[0]
    new_arg = astroid.nodes.AssignName('new_arg', parent=init_function.args)
    init_function.args.args.insert(1, new_arg)


# MRE
def class_wrapper(cls):
    """
    Class decorator which adds an extra argument to __init__
    """
    init_func = cls.__init__

    def wrapped_init(self, new_arg, *args, **kwargs):
        print(new_arg)
        init_func(self, *args, **kwargs)
    cls.__init__ = wrapped_init
    return cls


@class_wrapper
class CoolClass():
    def __init__(self):
        print('__init__')


if __name__ == '__main__':
    instance = CoolClass('123')
else:
    astroid.MANAGER.register_transform(astroid.ClassDef, add_extra_arg, has_extra_arg_decorator)
    print('plugin registered!')

```
You get the following error when running pylint without plugins:
```console
> python -m  pylint -E mwe.py
************* Module mwe
mwe.py:42:15: E1121: Too many positional arguments for constructor call (too-many-function-args)
```
This is most likely since it can't figure out what the decorator is doing (similar to #259 but for the class decorator).

If we enable the plugin which adds the arg back using the transform (since we can't use `signature-mutators` as it doesn't seem to work as expected here) we see that the error disappears
```console
> python -m  pylint -E --init-hook="import sys; sys.path.insert(0, '.')" --load-plugins mwe  mwe.py
plugin registered!
transform triggered!
```
If we however increase the number of jobs, we see that plugin is registered, but the transform not triggered, resulting in the error returning
```console
> python -m  pylint -j2 -E --init-hook="import sys; sys.path.insert(0, '.')" --load-plugins mwe  mwe.py
plugin registered!
************* Module mwe
mwe.py:42:15: E1121: Too many positional arguments for constructor call (too-many-function-args)
```

# Related issues
* #374
* #2573
* #4356

For the record, this bug specifically happens on Pylint >= 2.5 on a platform for which `multiprocessing.get_start_method != 'fork'` (such as [macOS on Python 3.8+](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.get_start_method))

```
$ python -c 'import multiprocessing as mp, platform; print("mp start method: " + mp.get_start_method()); print("platform: " + platform.system()); print("python: " + platform.python_version())'
mp start method: spawn
platform: Darwin
python: 3.8.10
$ pip install 'pylint==2.12.2'
$ pylint -j2 -E --init-hook="import sys; sys.path.insert(0, '.')" --load-plugins mwe mwe.py
plugin registered!
************* Module mwe
mwe.py:42:15: E1121: Too many positional arguments for constructor call (too-many-function-args)
```

Note that this works fine on versions of Pylint < 2.5. With the example above:

```
$ python -c 'import multiprocessing as mp, platform; print("mp start method: " + mp.get_start_method()); print("platform: " + platform.system()); print("python: " + platform.python_version())'
mp start method: spawn
platform: Darwin
python: 3.8.10
$ pip install 'pylint==2.4.4'
$ pylint -j2 -E --init-hook="import sys; sys.path.insert(0, '.')" --load-plugins mwe mwe.py
plugin registered!
plugin registered!
transform triggered!
```

Also note that things work fine on a platform such as Linux:

```
$ python -c 'import multiprocessing as mp, platform; print("mp start method: " + mp.get_start_method()); print("platform: " + platform.system()); print("python: " + platform.python_version())'
mp start method: fork
platform: Linux
python: 3.8.10
$ pip install 'pylint==2.12.2'
$ pylint -j2 -E --init-hook="import sys; sys.path.insert(0, '.')" --load-plugins mwe mwe.py
plugin registered!
transform triggered!
```

It looks like this bug was introduced by this big refactor: https://github.com/PyCQA/pylint/pull/3016 (you can see some plugin loading code getting removed [here](https://github.com/PyCQA/pylint/pull/3016/files#diff-fefd7a80fbed9f2ba1ad6fe5c19ecb0faebaf58df09ee35be231f8c74bc8529eL275)), which I believe went out in Pylint 2.5. That change relies upon the (IMO, incorrect) assumption that the only side effects of Pylint plugins are to alter the `PyLinter` object (which could get pickled and then unpickled in the subprocess), but some Pylint plugins actually register astroid transforms, which get lost when spawning a subprocess (but they are preserved when forking). I personally ran into this in a different way: we've been working around [PyCQA/astroid#786 with a custom plugin that calls `sys.setrecursionlimit(...)`](https://github.com/PyCQA/astroid/issues/786#issuecomment-716662442), and that change is now longer affecting subprocesses.

For the record, it was really hard for me to tell if this is a known bug or not.

https://pylint.pycqa.org/en/latest/user_guide/run.html#parallel-execution says:
   > There are some limitations in running checks in parallel in the current implementation. It is not possible to use custom plugins (i.e. --load-plugins option), nor it is not possible to use initialization hooks (i.e. the --init-hook option).

So it sort of sounds like this is a known issue, but then I went digging to see if there's a corresponding issue, and I learned that the history is a little more interesting than that:

1. (2014-10) The original implementation of `--jobs` adds that documentation saying `--load-plugins` and `--init-hook` don't work: https://github.com/PyCQA/pylint/commit/c03fefdefba97e187f5458288227f4224ee53cc0
2. (2015-05) Support for `--load-plugins` with `--jobs` is added! https://github.com/PyCQA/pylint/commit/0cf4153a638be20dff421bc9b890ffa30aeb40d3. The docs are not updated, though, so there's still a message saying `--load-plugins` does not work.
3. (2019-10) There's this big refactor of `--jobs`: https://github.com/PyCQA/pylint/pull/3016, which introduces the issue I've described above. So ironically, the docs are sort of up to date again.

--- a/doc/user_guide/usage/run.rst
+++ b/doc/user_guide/usage/run.rst
@@ -160,10 +160,10 @@ This will spawn 4 parallel Pylint sub-process, where each provided module will
 be checked in parallel. Discovered problems by checkers are not displayed
 immediately. They are shown just after checking a module is complete.
 
-There are some limitations in running checks in parallel in the current
-implementation. It is not possible to use custom plugins
-(i.e. ``--load-plugins`` option), nor it is not possible to use
-initialization hooks (i.e. the ``--init-hook`` option).
+There is one known limitation with running checks in parallel as currently
+implemented. Since the division of files into worker processes is indeterminate,
+checkers that depend on comparing multiple files (e.g. ``cyclic-import``
+and ``duplicate-code``) can produce indeterminate results.
 
 Exit codes
 ----------

--- a/doc/whatsnew/fragments/4874.bugfix
+++ b/doc/whatsnew/fragments/4874.bugfix
@@ -1,0 +1,5 @@
+``--jobs`` can now be used with ``--load-plugins``.
+
+This had regressed in astroid 2.5.0.
+
+Closes #4874

--- a/pylint/lint/parallel.py
+++ b/pylint/lint/parallel.py
@@ -52,6 +52,11 @@ def _worker_initialize(
     _worker_linter.set_reporter(reporters.CollectingReporter())
     _worker_linter.open()
 
+    # Re-register dynamic plugins, since the pool does not have access to the
+    # astroid module that existed when the linter was pickled.
+    _worker_linter.load_plugin_modules(_worker_linter._dynamic_plugins, force=True)
+    _worker_linter.load_plugin_configuration()
+
     if extra_packages_paths:
         _augment_sys_path(extra_packages_paths)
 

--- a/pylint/lint/pylinter.py
+++ b/pylint/lint/pylinter.py
@@ -363,15 +363,18 @@ class PyLinter(
         checkers.initialize(self)
         reporters.initialize(self)
 
-    def load_plugin_modules(self, modnames: list[str]) -> None:
+    def load_plugin_modules(self, modnames: list[str], force: bool = False) -> None:
         """Check a list of pylint plugins modules, load and register them.
 
         If a module cannot be loaded, never try to load it again and instead
         store the error message for later use in ``load_plugin_configuration``
         below.
+
+        If `force` is True (useful when multiprocessing), then the plugin is
+        reloaded regardless if an entry exists in self._dynamic_plugins.
         """
         for modname in modnames:
-            if modname in self._dynamic_plugins:
+            if modname in self._dynamic_plugins and not force:
                 continue
             try:
                 module = astroid.modutils.load_module_from_name(modname)

--- a/pylint/lint/pylinter.py
+++ b/pylint/lint/pylinter.py
@@ -13,7 +13,7 @@ import sys
 import tokenize
 import traceback
 from collections import defaultdict
-from collections.abc import Callable, Iterator, Sequence
+from collections.abc import Callable, Iterable, Iterator, Sequence
 from io import TextIOWrapper
 from pathlib import Path
 from re import Pattern

--- a/pylint/lint/pylinter.py
+++ b/pylint/lint/pylinter.py
@@ -18,7 +18,7 @@ from io import TextIOWrapper
 from pathlib import Path
 from re import Pattern
 from types import ModuleType
-from typing import Any, Protocol
+from typing import Any, Iterable, Protocol
 
 import astroid
 from astroid import nodes
@@ -363,7 +363,7 @@ class PyLinter(
         checkers.initialize(self)
         reporters.initialize(self)
 
-    def load_plugin_modules(self, modnames: list[str], force: bool = False) -> None:
+    def load_plugin_modules(self, modnames: Iterable[str], force: bool = False) -> None:
         """Check a list of pylint plugins modules, load and register them.
 
         If a module cannot be loaded, never try to load it again and instead

--- a/pylint/lint/pylinter.py
+++ b/pylint/lint/pylinter.py
@@ -18,7 +18,7 @@ from io import TextIOWrapper
 from pathlib import Path
 from re import Pattern
 from types import ModuleType
-from typing import Any, Iterable, Protocol
+from typing import Any, Protocol
 
 import astroid
 from astroid import nodes

--- a/tests/test_check_parallel.py
+++ b/tests/test_check_parallel.py
@@ -14,10 +14,11 @@ import sys
 from concurrent.futures import ProcessPoolExecutor
 from concurrent.futures.process import BrokenProcessPool
 from pickle import PickleError
+from typing import TYPE_CHECKING
+from unittest.mock import patch
 
 import dill
 import pytest
-from astroid import nodes
 
 import pylint.interfaces
 import pylint.lint.parallel
@@ -29,6 +30,9 @@ from pylint.lint.parallel import check_parallel
 from pylint.testutils import GenericTestReporter as Reporter
 from pylint.typing import FileItem
 from pylint.utils import LinterStats, ModuleStats
+
+if TYPE_CHECKING:
+    from astroid import nodes
 
 
 def _gen_file_data(idx: int = 0) -> FileItem:
@@ -181,6 +185,17 @@ class TestCheckParallelFramework:
                 linter=dill.dumps(linter), extra_packages_paths=["fake-path"]
             )
             assert "fake-path" in sys.path
+
+    def test_worker_initialize_reregisters_custom_plugins(self) -> None:
+        linter = PyLinter(reporter=Reporter())
+        linter.load_plugin_modules(["pylint.extensions.private_import"])
+
+        pickled = dill.dumps(linter)
+        with patch(
+            "pylint.extensions.private_import.register", side_effect=AssertionError
+        ):
+            with pytest.raises(AssertionError):
+                worker_initialize(linter=pickled)
 
     @pytest.mark.needs_two_cores
     def test_worker_initialize_pickling(self) -> None:

